File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> ret(nums.size(), 1);

        int left = 1;
        for (int i = 0; i < nums.size(); i++) {
            ret[i] *= left;
            left *= nums[i];
        }

        int right = 1;
        for (int i = nums.size() -1; i >= 0; i--) {
            ret[i] *= right;
            right *= nums[i];
        }

        return ret;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        sz: int = len(nums)
        ret: List[int] = [1] * sz

        left: int = 1
        for i in range(sz):
            ret[i] *= left
            left *= nums[i]

        right: int = 1
        for i in range(sz-1, -1, -1):
            ret[i] *= right
            right *= nums[i]
        
        return ret
